services/src/main/java/org/keycloak/crypto/Aes128GcmEncryptor.java

```diff
@@ -0,0 +1,39 @@
+package org.keycloak.crypto;
+
+import javax.crypto.Cipher;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+import java.util.UUID;
+
+public class Aes128GcmEncryptor {
+    private static final int IV_SIZE_BYTE = 12;
+    private static final int AUTH_TAG_SIZE_BYTE = 16;
+    private static final String ALGORITHM = "AES/GCM/NoPadding";
+    private static final byte[] IV = new byte[IV_SIZE_BYTE];
+
+    private Aes128GcmEncryptor() {
+    }
+
+    public static byte[] encrypt(String secret, byte[] bytes) throws GeneralSecurityException {
+        return getCipher(Cipher.ENCRYPT_MODE, secret).doFinal(bytes);
+    }
+
+    public static byte[] decrypt(String secret, byte[] encryptedBytes) throws GeneralSecurityException {
+        return getCipher(Cipher.DECRYPT_MODE, secret).doFinal(encryptedBytes);
+    }
+
+    private static Cipher getCipher(int mode, String secret) throws GeneralSecurityException {
+        UUID uuid = UUID.fromString(secret);
+        byte[] bytes = ByteBuffer.wrap(new byte[AUTH_TAG_SIZE_BYTE])
```

Comment:
I'm not sure whether AAD is required or not in this case.
Is it safe to encrypt/decrypt data in AES-GCM without AAD in this case?

Reply:
In general it should not influence security of the encryption/decryption process greatly.
As I've seen in documentation if AAD is used it should be re-generated for each pair encrypt/decrypt operation, but then we will need to store it somewhere, like in existing code ADD of JWE it is stored in base64Header of JWE, here I did not find a place to store or get the AAD from in our flow, that is why I simply mocked it, we could add corresponding bytes to the. beginning or the end of the login_hint, but that would mean anybody can get it from there and there actually is no reason to have it as such. Any suggestion from your side?

Reply:
And AAD is required for decryption by the Cipher.init(), so can't simply remove it.

Reply:
I see, in this case, IV and AAD can not be shared among the client and keycloak by `login_hint` itself unlike JWE whose format includes IV and AAD. Thank you.

Reply:
I see, in this case, IV and AAD can not be shared among the client and keycloak by `login_hint` itself unlike JWE whose format includes IV and AAD. Thank you.